File: LeetCodeProblems/LeetCodeProblems-Swift/LeetCodeProblems-Swift/BacktrackProblems/Combine.swift
//
//  Combine.swift
//  LeetCodeProblems-Swift
//
//  Created by Ficow on 2021/7/16.
//

import Foundation

final class Combine {
    func run() {
        testCombine()
    }

    func testCombine() {
        func judge(n: Int, k: Int, expected: [[Int]]) {
            printAndAssert(result: Set(combine111(n, k).map { $0 }),
                           expected: Set(expected.map { $0 }))
        }
        judge(n: 4, k: 2,
              expected: [
                [2,4],
                [3,4],
                [2,3],
                [1,2],
                [1,3],
                [1,4],
              ])
        judge(n: 1, k: 1, expected: [[1]])
        judge(n: 4, k: 4, expected: [[1, 2, 3, 4]])
        judge(n: 4, k: 3, expected: [
            [1, 2, 3],
            [1, 2, 4],
            [1, 3, 4],
            [2, 3, 4],
        ])
    }















    func combine2(_ n: Int, _ k: Int) -> [[Int]] {
        var result = [[Int]]()
        func append(_ i: Int, to list: inout [Int]) {
            if list.count == k {
                result.append(list)
                return
            }
            // with (k > list.count + n - j + 1), this `if` is useless
//            if i > n {
//                return
//            }
            for j in i...n {
                if k > list.count + n - j + 1 {
                    return
                }
                list.append(j)
                append(j+1, to: &list)
                _ = list.popLast()
            }
        }
        var list = [Int]()
        append(1, to: &list)
        return result
    }

































    func combine1(_ n: Int, _ k: Int) -> [[Int]] {
        var result = [[Int]]()
        func push(index: Int, list: inout [Int]) {
            if list.count == k {
                result.append(list)
                //                print(list)
                return
            }
            // with (k - list.count > n - i + 1), this `if` is useless
//            if index > n {
//                //                print(">")
//                return
//            }
            //            print(list)
            for i in index...n {
                // cannot get k elements
                if k - list.count > n - i + 1 {
                    return
                }
                list.append(i)
                push(index: i+1, list: &list)
                _ = list.popLast()
            }
        }
        var list = [Int]()
        push(index: 1, list: &list)
        return result
    }

    // https://leetcode.com/problems/combinations/discuss/26992/Short-Iterative-C%2B%2B-Answer-8ms
    func combine11(_ n: Int, _ k: Int) -> [[Int]] {
        var result = [[Int]]()
        var nums = [Int](repeating: 0, count: k)
        var i = 0
        while i >= 0 {
            nums[i] += 1
            // n=4,k=2, [1,2,3,4] -> [2,2,3,4] -> terminate
            if n - nums[i] + i + 1 < k {
//          if nums[i] > n {
//                print(nums)
                i -= 1
            } else if i == k - 1 {
                result.append(nums)
            } else {
                i += 1
                nums[i] = nums[i - 1]
            }
        }
//        print("---")
        return result
    }

    func combine111(_ n: Int, _ k: Int) -> [[Int]] {
        // n=4, k=3
        // 1, 2, 3, 5
        //       ^
        // 1, 2, 4, 5
        //    ^
        // 1, 3, 4, 5
        //    ^
        // 2, 3, 4, 5
        // ^
        // ----------
        // 1, 3, 4, 5
        //    ^
        // 1, 2, 4, 5
        //       ^
        // 1, 2, 3, 5
        //          ^
        // ----------
        // 1, 2, 3, 6
        //          ^
        var ans = [[Int]]()
        var temp = [Int]()
        for i in 1...k {
            temp.append(i)
        }
        // allow j increase to k
        temp.append(n+1)
        var j = 0
        while j < k {
            ans.append(Array(temp[0...k-1]))
            j = 0
            // find the gap
            while j < k, temp[j] + 1 == temp[j + 1] {
                temp[j] = j + 1
                j += 1
            }
            // increase the last gap to meet n
            temp[j] += 1
//            print(temp)
        }
        return ans
    }
}

File: LeetCodeProblems/LeetCodeProblems-C++/LeetCodeProblems-C++/BacktrackProblems/Combine.cpp
//
//  Combine.cpp
//  LeetCodeProblems-C++
//
//  Created by Ficow on 2021/7/17.
//

#include "Combine.hpp"

vector<vector<int>> combine(int n, int k);

bool equalVectors(vector<vector<int>> v1, vector<vector<int>> v2) {
    // https://stackoverflow.com/a/5225902
    // https://stackoverflow.com/a/284004
    set<vector<int>> s1(v1.begin(), v1.end());
    set<vector<int>> s2(v2.begin(), v2.end());
    set<vector<int>> result;
    set_difference(s1.begin(), s1.end(), s2.begin(), s2.end(), inserter(result, result.end()));
    if (!result.empty()) {
        return false;
    }
    result.clear();
    set_difference(s2.begin(), s2.end(), s1.begin(), s1.end(), inserter(result, result.end()));
    return result.empty();
}

void printVector(vector<int>& v) {
    cout << "[";
//    for (auto it = v.begin(); it != v.end(); ++it) {
//        cout << *it << ",";
//    }
    for (auto i = 0; i < v.size(); ++i) {
        cout << v[i] << ",";
    }
    cout << "]" << endl;
}


vector<vector<int>> combine1(int n, int k) {
    vector<int> temp;
    vector<vector<int>> ans;
    // 初始化
    // 将 temp 中 [0, k - 1] 每个位置 i 设置为 i + 1，即 [0, k - 1] 存 [1, k]
    // 末尾加一位 n + 1 作为哨兵
    for (int i = 1; i <= k; ++i) {
        temp.push_back(i);
    }
    temp.push_back(n + 1);
//    printVector(temp);

    int j = 0;
    while (j < k) {
        ans.emplace_back(temp.begin(), temp.begin() + k);
//        cout << endl << "-----------      append" << endl;
//        printVector(ans.back());
        j = 0;
        // 寻找第一个 temp[j] + 1 != temp[j + 1] 的位置 t
        // 我们需要把 [0, t - 1] 区间内的每个位置重置成 [1, t]
        if (j < k && temp[j] + 1 == temp[j + 1]) {
//            cout << endl << "-      while" << endl;
        }
        // 找到不连续的位置，然后增大
        while (j < k && temp[j] + 1 == temp[j + 1]) {
            temp[j] = j + 1;
//            printVector(temp);
            ++j;
//            cout << "j = " << j << endl;
        }
//        cout << "j = " << j << endl;
        // j 是第一个 temp[j] + 1 != temp[j + 1] 的位置
        ++temp[j];
//        cout << endl << "-      increase j:" << j << endl;
//        printVector(temp);
    }
    return ans;
}

void judge(int n, int k, vector<vector<int>> expected) {
    assert(equalVectors(combine1(n, k), expected));
}

void Combine::run() {
    judge(1, 1, vector<vector<int>>{vector<int>{1}});
    judge(4, 4, vector<vector<int>>{vector<int>{1, 2, 3, 4}});
    judge(4, 2, vector<vector<int>>{
        vector<int>{1, 2},
        vector<int>{1, 3},
        vector<int>{1, 4},
        vector<int>{2, 3},
        vector<int>{2, 4},
        vector<int>{3, 4},
    });
    judge(4, 3, vector<vector<int>>{
        vector<int>{1, 2, 3},
        vector<int>{1, 2, 4},
        vector<int>{1, 3, 4},
        vector<int>{2, 3, 4},
    });
}


void append(int i, int n, int k, vector<int>& list, vector<vector<int>>& result) {
    if (list.size() == k) {
        result.push_back(list);
        return;
    }
    if (i > n) {
        return;
    }
    for (int j = i; j <= n; j++) {
        if (k > list.size() + n - j + 1) {
            return;
        }
        list.push_back(j);
        append(j+1, n, k, list, result);
        list.pop_back();
    }
}

vector<vector<int>> combine(int n, int k) {
    vector<vector<int>> result;
    vector<int> list;
    append(1, n, k, list, result);
    return result;
}
